### PR TITLE
Add overloaded integer literals

### PR DESCRIPTION
**Pull request by eliaslfox**

### CODE REVIEW

**Comment by stylewarning** [comment]
This change definitely motivates constant folding.

File: src/ast/parse-form.lisp [hunk 1]
```diff
@@ -232,7 +232,14 @@ This does not attempt to do any sort of analysis whatsoever. It is suitable for
        (pattern-constructor ctor (mapcar #'parse-pattern args))))))
 
 (defun parse-atom (atom)
-  (node-literal atom atom))
+  ;; Convert interger literals into fromInt calls. This allowes for
```
Comment (stylewarning) [inline src/ast/parse-form.lisp:235]: Little spelling nits.

File: src/typechecker/predicate.lisp [hunk 2]
```diff
@@ -28,6 +28,10 @@
   "A list of type predicates"
   `(satisfies ty-predicate-list-p))
 
+(defun static-predicate-p (pred)
+  "Is PRED a static predicate (no type variables)"
+  (= 0 (length (coalton-impl/typechecker::type-variables (ty-predicate-types pred)))))
```
Comment (stylewarning) [inline src/typechecker/predicate.lisp:33]: zerop 

or if it's a list, just `endp` without length.

File: tests/free-variables-tests.lisp [hunk 3]
```diff
@@ -3,6 +3,7 @@
 (in-package #:coalton-tests)
 
 (deftest test-free-variables ()
```
Comment (stylewarning) [inline tests/free-variables-tests.lisp:5]: this can be deleted when other PR merges

File: src/typechecker/derive-type.lisp [hunk 4]
```diff
@@ -548,6 +548,16 @@ EXPL-DECLARATIONS is a HASH-TABLE from SYMBOL to SCHEME"
                      (some (lambda (b) (not (coalton-impl/ast::node-abstraction-p (cdr b))))
                            bindings)))
 
+            ;; NOTE: This is where defaulting happens
```
Comment (stylewarning) [inline src/typechecker/derive-type.lisp:551]: could you explain a little more in this comment how defaulting relates to the code below?

**Review by stylewarning — CHANGES_REQUESTED** [review]

**Review by stylewarning — CHANGES_REQUESTED** [review]
Code looks good. Last request, can you write a little documentation (maybe in the intro) about overloading and defaulting rules, with an example or two?

**Comment by eliaslfox** [comment]
Also it wouldn't be hard to add this for floats. We just need a type class similar to [fractional](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#t:Fractional).

**Review by stylewarning — APPROVED** [review]

**Pull request merged**